src/dotty/tools/dotc/typer/Checking.scala

```diff
@@ -57,6 +58,12 @@ object Checking {
           val bounds = tparams.map(tparam =>
             tparam.info.asSeenFrom(tycon.tpe.normalizedPrefix, tparam.owner.owner).bounds)
           checkBounds(args, bounds, _.substDealias(tparams, _))
+        case Select(qual, name) if name.isTypeName =>
+          checkRealizable(qual.tpe, qual.pos)
+        case SelectFromTypeTree(qual, name) if name.isTypeName =>
+          checkRealizable(qual.tpe, qual.pos)
+        case SingletonTypeTree(ref) =>
+          checkRealizable(ref.tpe, ref.pos)
         case _ =>
```

Comment:
What about `AndTypeTree`, `OrTypeTree, `RefinedTypeTree`, etc? Maybe we should add a case like:
```scala
case typ: TypTree =>
  checkRealizable(typ.tpe, typ.pos)
```
I think that if we do this, then my proposition at https://github.com/lampepfl/dotty/pull/1051#issuecomment-181572666 becomes safe, because `Foo[T & U]` will be invalid.